Migrating away from AngularJS

AmitMY opened the pull request:
Removing the dependency in AngularJS, moving to custom elements in ES6 and SCSS, more tests, tests coverage, better files structure.

`BLOG.md` files is the work in progress blog post to release after merging. It contains all the information on what was done.

Fixes: 
https://github.com/Webiks/force-horse/issues/2 - Adds a `readyEvent` Event emitter, whenever it redraws (new SVG)
https://github.com/Webiks/force-horse/issues/4 - Using an event emitter, can `unsubscribe`

Like everything, this needs many many many more tests, as the main provider is not fully covered.

Commit:
chore(): moving to the main repo

AmitMY committed:
tests(): add tests for event emitter and buttons

AmitMY committed:
tests(): add tests for buttons and main component

AmitMY committed:
tests(): add tests for component and provider

AmitMY committed:
chore(tests): add tests coverage

AmitMY committed:
docs(blog): add a blog

AmitMY committed:
fix(coverage): tests coverage works

AmitMY committed:
fix(buttons): add mask no repeat

AmitMY committed:
refactor(): remove the change for options, add a not-DOM intensive setOptions method

AmitMY committed:
feat(travis): add travis for release and npm

AmitMY committed:
fix(travis): use ci script

AmitMY committed:
docs(blog): tiny edit

AmitMY committed:
docs(blog): fix footer

AmitMY committed:
docs(README): update to current usage

AmitMY committed:
docs(README): fix formatting

AmitMY committed:
docs(README): fix formatting

AmitMY committed:
docs(): general info

AmitMY committed:
fix(component): fix attribute binding

ramtob reviewed (changes requested):
Great job. I entered several questions, comments, and requests for small changes.

AmitMY commented on BLOG.md:
```diff
@@ -0,0 +1,49 @@
+# Revisiting Force-Horse: Making a Legacy Plugin Agnostic and Usable
```
Thanks!
It is legacy, just because AngularJS is legacy, see:
> Legacy code refers to an application system source code type that is no longer supported

Yeah, you kind of mixed some ES6 in there, just a bit.

AmitMY committed:
docs(BLOG): AngularJS is not deprecated

AmitMY commented on BLOG.md:
```diff
@@ -0,0 +1,34 @@
+While you can read all of the details on the library in Ram's [blog post](http://webiks.com/force-horse-force-layout-component/), the gist of it is having a library that adapts to many types of data, is easily modifiable for developers, supports large data sets, and easily allowing customization of the base SVG's design.
+
+## So... What's New?
+Having the library written using AngularJS, forces the user to use a deprecated framework. Either the user thinks they can just include AngularJS in the bundle, increasing the bundle, or trying to implement the component for their used framework.
```
You are right. Changed from "deprecated" to "legacy".

AmitMY committed:
docs(BLOG): use custom elements v1

AmitMY commented on BLOG.md:
```diff
@@ -0,0 +1,34 @@
+Therefore, we decided to remove the dependency for AngularJS, making the library framework agnostic.
+
+##### Agnostification
+We migrated AngularJS out of the library, moving to a standard browser feature - [Custom Elements](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Custom_Elements), supported by most modern browsers, or being worked on at this time (In case it isn't supported, there is a tiny [browser polyfill](https://github.com/webcomponents/webcomponentsjs) - see [browsers support](https://caniuse. ...
```
1. Done
2. Done
3. We can, we shan't imo. It is super simple to add, I can add the code to the README, it is like:
- `npm i @webcomponents/custom-elements`
- Add `<script src="custom-elements.min.js"></script>` or `import from '@webcomponents/custom-elements/custom-elements.min.js'`

The reason I think we shan't is that if you use Angular or React or whatever, it is just bloat.

AmitMY commented on README.md:
````diff
@@ -68,38 +100,37 @@ The config parameters are:
 ```
 
-The forceHorse project contains a complete **demo application**. The demo application is also available on the [plunker](http://embed.plnkr.co/SYmehtaAnQVyMpLJJY2B/?show=preview) site.
+The forceHorse project contains a complete **demo application**. The demo application is also available on the [plunker](http://embed.plnkr.co/SYmehtaAnQVyMpLJJY2B/?show=preview) site.
````
Hahaha probably.
I'll make a new one on the updated code base, as the package is on NPM already (had to test it for the demo)

AmitMY commented on src/components/force-horse/force-horse.js:
```diff
@@ -0,0 +1,63 @@
+import {ForceHorseProvider} from '../../providers/force-horse';
```
Reasonable. Done

AmitMY commented on src/providers/force-horse.js:
```diff
@@ -0,0 +1,1220 @@
+/**
+ * Produces a class-instance for each instance of ForceHorse on a page
+ */
+export class ForceHorseProvider {
```
Because it does not instantiate a service, and just provides logic for the component. Can change if you find this awful

AmitMY commented on webpack.config.js:
```diff
@@ -0,0 +1,57 @@
+        use: extractSASS.extract({ // Instance 2
+          use: [
+            'css-loader',
+            'postcss-loader',
```
because I use scss, and because I want autoprefixer for `-webkit-mask..`, so `sass-loader` for scss, and `postcss` for `autoprefixer`

AmitMY commented on .travis.yml:
```diff
@@ -0,0 +1,23 @@
+language: node_js
```
Every commit, travis will automatically try to tests. and report back on github what is the status
If there is a new tag, it will deploy it to a github release, and to an NPM release, hopefully

AmitMY replied:
See below, it shows a PR status

AmitMY committed:
refactor(names): rename files

AmitMY committed:
refactor(): change options to config and data

AmitMY committed:
chore(): bump version

ramtob commented:
It looks fine, I'm just still a bit uncomfortable with the word "provider"... Did you see the word "provider" used in this way anywhere else? I know the word is used in AngularJS, but it has a different meaning there (a kind of service generator). What do you think about "viewer" instead? Because what you called the force horse "component" is actually a container, wrapping the graph viewer (what you called "provider") and the buttons bars.

AmitMY committed:
refactor(): rename force-horse-viewer

AmitMY commented:
@ramtob Ok, done.
I also moved `data` and `config` to be attributes instead of options, as it makes more sense in an operative stand point, and added Angular integration in the README

AmitMY committed:
fix(): moved to viewer

ramtob commented:
Good. Just one more little request: instead of forceHorse.instance and this.instance, call it forceHorce.viewer and this.viewer. This will be more informative and readable (this.instance is actually a kind of duplication, since "this" is already an indication for an instance)

AmitMY committed:
refactor(instance): change to viewer

AmitMY commented:
@ramtob Fair point, changed that.
Plunker/Demo apps will be up after your PR approval

ramtob commented:
I approved the PR. Nice work

AmitMY commented:
Thanks!
